daemon/oci_linux.go

```diff
@@ -818,8 +835,19 @@ func WithCgroups(daemon *Daemon, c *container.Container) coci.SpecOpts {
 			parentPath = filepath.Clean("/" + parentPath)
 		}
 
-		if err := daemon.initCgroupsPath(parentPath); err != nil {
-			return fmt.Errorf("linux init cgroups path: %v", err)
+		mnt, root, err := cgroups.FindCgroupMountpointAndRoot("", "cpu")
+		if err != nil {
+			return errors.Wrap(err, "unable to init CPU RT controller")
+		}
+		// When docker is run inside docker, the root is based of the host cgroup.
+		// Should this be handled in runc/libcontainer/cgroups ?
+		if strings.HasPrefix(root, "/docker/") {
+			root = "/"
```

Comment:
not new here, but wouldn't it work with systemd?

Reply:
This is a dirty hack in here, and I didn't want to mess with it as it's easy to break things.

Most probably yes, it won't work since the prefix would be `system.slice` or so.

To be honest, even after looking at all this code for a few montjs, I still can't figure out why mountinfo's `root` field is needed anywhere :( and this missing knowledge prevents me from optimizing runc :(

An alternative to the above would be to not use `root` field from mountinfo, i.e.

```go
               mnt, err := cgroups.FindCgroupMountpoint("", "cpu")
               if err != nil {
                       return errors.Wrap(err, "unable to init CPU RT controller")
               }
               if err := daemon.initCpuRtController(mnt, parentPath); err != nil {
                       return errors.Wrap(err, "unable to init CPU RT controller")
```

but I am hesitant to remove it as I don't fully understand why it's needed here.

@AkihiroSuda maybe you can shed some light on it?

Reply:
Here's a similar code in runc:
https://github.com/opencontainers/runc/blob/7673bee6bfbc28c8cfbc64165eea29b83e9957f0/libcontainer/cgroups/utils.go#L338-L352

The code initially comes from https://github.com/opencontainers/runc/pull/135, which explains that it is needed for the docker-in-docker case.

In this case, I think, we won't be able to apply CPU RT settings as we don't have access to a parent cgroup, so it does not make sense here.

Reply:
@AkihiroSuda @kolyshkin still feedback needed on this one? ^^

Reply:
@thaJeztah the concern raised by @AkihiroSuda is valid, but this is something that is better to be addressed separately as a followup, otherwise we're putting too many eggs in one basket.